vendor/knative.dev/pkg/controller/controller.go

```diff
@@ -273,7 +273,12 @@ func (c *Impl) Run(threadiness int, stopCh <-chan struct{}) error {
 	defer runtime.HandleCrash()
 	sg := sync.WaitGroup{}
 	defer sg.Wait()
-	defer c.WorkQueue.ShutDown()
+	defer func() {
+		c.WorkQueue.ShutDown()
+		for c.WorkQueue.Len() > 0 {
```

Comment:
I have somewhat of a hard time understanding why this changes the semantics at all.

The `sg.Wait` above waits until the `processNextWorkItem` loops finish, which is only the case if `processNextWorkItem` returns false. It only does so if `WorkQueue.Get()` returns `shutdown = true` which it'll only do if `len(q.queue) == 0`, which seems to be exactly what we're replicating here.

https://github.com/kubernetes/client-go/blob/4f3abb12cae2d6817e1b6d72d9cbc3d3a6ddf965/util/workqueue/queue.go#L126-L129

Reply:
Right, so after `Shutdown` returns, we wait for `sg.Wait()` which has the same exit criteria basically as the for loop shown here.

Reply:
In effect :)